keel-sql/src/main/resources/db/changelog/20200714-add-event-application-column.yml

```diff
@@ -0,0 +1,28 @@
+databaseChangeLog:
+- changeSet:
+    id: add-event-application-column
+    author: fletch
+    changes:
+    - sql:
+        sql: |
+          alter table event add column application varchar(255) generated always as (json ->> '$.application') stored;
+    - createIndex:
+        tableName: event
+        indexName: event_ref_scope_application_timestamp_idx
+        columns:
```

Comment:
If I'm reading the where clause of the `eventHistory` query correctly, it looks like this:

```sql
select json from event
where 
  (scope="resource" and ref=uid) or
  (scope="application" and application=application)
...
```
This index is defined as `(ref,scope,application,timestamp)`.

I think the event history query will still require a full table scan, because the second part of the where clause can't take full advantage of the index. I think you need to define two indexes to avoid the table scan:

1. `(scope,ref,timestamp)` 
2. `(scope,application,timestamp)`

Hopefully, someone with more database experience than I can chime in here.

Reply:
That's how _I_ always thought indices worked but @marchello2000 suggested that queries can benefit from using a subset of the columns in an index. Would be interested in his take on this.

Reply:
For the index to be used (other than query optimizer deciding to use it, etc...) the columns used in the where clause have to be "contiguous" in the index. so if index is `col1, col2, col3` and you do a where clause for `col1=x or col3=y` you will get the benefit of index for `col1=x` but not for `col3=y`

So I think here you'd want to the `scope` to be first column of the index since it's used in both parts of the `or`. But after that only one of them will benefit from the index, so pick the one that has the most cardinality

you can make sure it uses the index with
```sql
explain format=json select json from event...
```

Reply:
i am not sure what happens with the `or` it (MySQL) might just decide full table scan is more efficient.. You'd have to try out - I don't know for sure

Reply:
I think it's worth doing a mini-benchmark test with one index and two indexes and see what the difference in performance is.

Reply:
With my original index: `"query_cost": "53658.80"`
With revised 2 indices: `"query_cost": "146.21"`

I think you were right @lorin !